### ISSUE TITLE
Local docker cluster doesn't populate service account token

### ISSUE THREAD

**Issue by mwielgus** [kubernetes/kubernetes#11984]
Local cluster (created with http://kubernetes.io/v1.0/docs/getting-started-guides/docker.html):

```
$ kubectl -s http://localhost:8080 exec cassandra -- ls /var/run/secrets
ls: cannot access /var/run/secrets: No such file or directory
error: Error executing remote command: Error executing command in container: Error executing in Docker Container: 2
```

GCE cluster:

```
$ kubectl exec cassandra -- ls /var/run/secrets
kubernetes.io
```

Examples that depend on Pod<->API communication crash because of this. Ref: #10931

**Comment by alex-mohr** [comment]
@eparis Do I remember that you did much of the secrets work?  If so, could you please take a look?

**Comment by smarterclayton** [comment]
If you are running Kubernetes (specifically the kubelet) in Docker you need to be running a Red Hat distro version of Docker which carries the SHARED mount propagation patch.  Otherwise, secrets and other volumes can't be mounted in pods and thus will not work.

**Comment by eparis** [comment]
not SHARED.  SLAVE.

**Comment by eparis** [comment]
But I think clayton is mostly right.  Upstream docker has turned off (in code, not a config option) functionality required to make it possible to run kubelet...

I'd love to see logs from the master and node containers though, to dig deeper.

**Comment by timothysc** [comment]
cross-ref: https://github.com/docker/docker/pull/14097#issuecomment-127301915

**Comment by mwielgus** [comment]
cc: @fgrzadkowski

**Comment by fgrzadkowski** [comment]
@eparis @smarterclayton 
Would it help if we manually change mount namespace for kubelet using ```mount --make-rshared``` command? I have to admit that this is completely unknown area for me, so it's just shooting in the dark :)

**Comment by smarterclayton** [comment]
Maybe?  I haven't tried it.

On Tue, Aug 25, 2015 at 8:32 AM, Filip Grzadkowski <notifications@github.com
> wrote:

> @eparis <https://github.com/eparis> @smarterclayton
> <https://github.com/smarterclayton>
> Would it help if we manually change mount namespace for kubelet using mount
> --make-rshared command? I have to admit that this is completely unknown
> area for me, so it's just shooting in the dark :)
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/kubernetes/kubernetes/issues/11984#issuecomment-134559808>
> .
>



-- 
Clayton Coleman | Lead Engineer, OpenShift

**Comment by errordeveloper** [comment]
@pmorie said:
> Now with more hacks. Tmpfs e2e runs locally with kubelet running in container; secrets doesn't. Next step will be to investigate secrets.

(https://github.com/kubernetes/kubernetes/pull/6936#issuecomment-94083734)

It's not clear from the whole #6936 thread whether that ever been investigated...

**Comment by errordeveloper** [comment]
I wonder if #6936  was done with the assumption that either `SHARED` or `SLAVE` propagation was enabled in Docker?

**Comment by smarterclayton** [comment]
RedHat distros ship Docker with SLAVE/SHARED (don't remember which,
but since 1.8), our original testing was was SLAVE, but later patches
fixed most of the known issues w.r.t. propagation hacks.  When 1.10
lands the mount options will be used to launch the kubelet to avoid
the need for the hacks.

**Comment by bobrik** [comment]
I think the issue is resolved in docker 1.12.

**Comment by kargakis** [comment]
Shouldn't be an issue anymore.
